src/laddie/laddie_velocity.f90

```diff
@@ -0,0 +1,420 @@
+
+    CALL gather_to_all_logical_1D( laddie%mask_a, mask_a_tot)
+
+    ! Initialise ambient T and S                             
+    ! TODO costly, see whether necessary to recompute with Hstar
+    CALL compute_ambient_TS( mesh, ice, ocean, laddie, Hstar)
+
+    ! Compute buoyancy
+    CALL compute_buoyancy( mesh, ice, laddie, npx, Hstar)
+ 
+    ! Bunch of mappings                                      
+    CALL map_a_b_2D( mesh, laddie%detr, detr_b)
+    CALL map_H_a_b( mesh, laddie, laddie%Hdrho_amb, laddie%Hdrho_amb_b)
+    CALL map_H_a_b( mesh, laddie, Hstar, Hstar_b)
+    CALL map_H_a_c( mesh, laddie, Hstar, Hstar_c)
+
+    ! Bunch of derivatives
+    CALL ddx_a_b_2D( mesh, laddie%drho_amb, laddie%ddrho_amb_dx_b)
+    CALL ddy_a_b_2D( mesh, laddie%drho_amb, laddie%ddrho_amb_dy_b)
+    CALL ddx_a_b_2D( mesh, Hstar, laddie%dH_dx_b)
+    CALL ddy_a_b_2D( mesh, Hstar, laddie%dH_dy_b)
+
+    ! Compute divergence of momentum
+    SELECT CASE(C%choice_laddie_momentum_advection)
+      CASE DEFAULT
+        CALL crash('unknown choice_laddie_momentum_advection "' // TRIM( C%choice_laddie_momentum_advection) // '"')
+      CASE ('none')
+        laddie%divQU = 0.0_dp
+        laddie%divQV = 0.0_dp
+      CASE ('upstream')
+        !CALL compute_divQUV_upstream( mesh, laddie, npx, Hstar_b)
+        CALL compute_divQUV_upstream( mesh, laddie, npx, npxref%H_b)
+    END SELECT
+
+    ! == Integrate U and V ==
+    ! =======================
+
+    ! Loop over vertices
+    DO ti = mesh%ti1, mesh%ti2
+      IF (laddie%mask_b( ti)) THEN
+
+        ! == pressure gradient force ==
+        ! =============================
+
+        IF (laddie%mask_cf_b( ti) .OR. laddie%mask_gl_b( ti)) THEN
+          ! Assume dH/dx and ddrho/dx = 0
+
+          ! Define PGF at calving front / grounding line
+          PGF_x = grav * laddie%Hdrho_amb_b( ti) * laddie%dHib_dx_b( ti) ! &
+
+          PGF_y = grav * laddie%Hdrho_amb_b( ti) * laddie%dHib_dy_b( ti) ! &
+        ELSE
+          ! Regular full expression
+          PGF_x = - grav * laddie%Hdrho_amb_b( ti) * laddie%dH_dx_b( ti) &
+                  + grav * laddie%Hdrho_amb_b( ti) * laddie%dHib_dx_b( ti) &
+                  - 0.5*grav * Hstar_b( ti)**2 * laddie%ddrho_amb_dx_b( ti)
+
+          PGF_y = - grav * laddie%Hdrho_amb_b( ti) * laddie%dH_dy_b( ti) &
+                  + grav * laddie%Hdrho_amb_b( ti) * laddie%dHib_dy_b( ti) &
+                  - 0.5*grav * Hstar_b( ti)**2 * laddie%ddrho_amb_dy_b( ti)
+        END IF
+
+        ! == time derivatives ==
+        ! ======================
+
+        ! dHU_dt
+        dHUdt = - laddie%divQU( ti) &
+                + PGF_x &
+                + C%uniform_laddie_coriolis_parameter * Hstar_b( ti) * npxref%V( ti) &
+                - C%laddie_drag_coefficient_mom * npxref%U( ti) * (npxref%U( ti)**2 + npxref%V( ti)**2)**.5 &
+                - detr_b( ti) * npxref%U( ti)
+
+        IF (inclvisc) THEN
+          dHUdt = dHUdt + laddie%viscU( ti)
```

Comment:
Yoooo this is what I meant earlier! :D Here too, please rename inclvisc to something more human-readable.

Reply:
Hehe will do